src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs

```diff
@@ -203,43 +222,56 @@ addVariable x = do
 type ConstraintMonomial = Mono Natural Natural
 
 -- | The type that represents a constraint in the arithmetic circuit.
-type Constraint c = Poly c Natural Natural
+type Constraint c i = Poly c (Var i) Natural
 
 -- | Adds a constraint to the arithmetic circuit.
-addConstraint :: Arithmetic a => Constraint a -> State (ArithmeticCircuit a U1) ()
+addConstraint :: Arithmetic a => Constraint a i -> State (ArithmeticCircuit a i U1) ()
 addConstraint c = zoom #acSystem . modify $ insert (toVar [] c) c
 
-rangeConstraint :: Natural -> a -> State (ArithmeticCircuit a U1) ()
+rangeConstraint :: Natural -> a -> State (ArithmeticCircuit a i U1) ()
 rangeConstraint i b = zoom #acRange . modify $ insert i b
 
 -- | Adds a new variable assignment to the arithmetic circuit.
 -- TODO: forbid reassignment of variables
-assignment :: Natural -> (Map Natural a -> a) -> State (ArithmeticCircuit a U1) ()
+assignment :: Natural -> (i a -> a) -> State (ArithmeticCircuit a i U1) ()
 assignment i f = zoom #acWitness . modify $ insert i f
 
 -- | Evaluates the arithmetic circuit with one output using the supplied input map.
-eval1 :: ArithmeticCircuit a Par1 -> Map Natural a -> a
-eval1 ctx i = witnessGenerator ctx i ! unPar1 (acOutput ctx)
+eval1 :: Representable i => ArithmeticCircuit a i Par1 -> i a -> a
+eval1 ctx i = case unPar1 (acOutput ctx) of
+    NewVar k -> witnessGenerator ctx i ! k
+    InVar j  -> index i j
 
 -- | Evaluates the arithmetic circuit using the supplied input map.
-eval :: Functor o => ArithmeticCircuit a o -> Map Natural a -> o a
-eval ctx i = (witnessGenerator ctx i !) <$> acOutput ctx
+eval :: (Representable i, Functor o) => ArithmeticCircuit a i o -> i a -> o a
+eval ctx i = acOutput ctx <&> \case
+    NewVar k -> witnessGenerator ctx i ! k
+    InVar j -> index i j
 
 -- | Evaluates the arithmetic circuit with no inputs and one output using the supplied input map.
-exec1 :: ArithmeticCircuit a Par1 -> a
-exec1 ac = eval1 ac empty
+exec1 :: ArithmeticCircuit a U1 Par1 -> a
+exec1 ac = eval1 ac U1
 
 -- | Evaluates the arithmetic circuit with no inputs using the supplied input map.
-exec :: Functor o => ArithmeticCircuit a o -> o a
-exec ac = eval ac empty
+exec :: Functor o => ArithmeticCircuit a U1 o -> o a
+exec ac = eval ac U1
 
 -- | Applies the values of the first `n` inputs to the arithmetic circuit.
 -- TODO: make this safe
-apply :: [a] -> State (ArithmeticCircuit a U1) ()
-apply xs = do
-    inputs <- gets acInput
-    zoom #acInput . put $ drop (length xs) inputs
-    zoom #acWitness . modify . union . fromList $ zip inputs (map const xs)
+apply :: (Eq a, Field a, Ord (Rep j), Scale a a, FromConstant a a, Representable i) => i a -> ArithmeticCircuit a (i :*: j) U1 -> ArithmeticCircuit a j U1
```

Comment:
Also `Scale a a` and `FromConstant a a` turned up all over the place. Might as well add them to `Arithmetic a` constraint in the end

Reply:
Discussed it in #154